File: cpp/pure.cpp
#include <cstdlib>
#include <iostream>


int power(int base, int exponent){
    if (exponent == 0){return 1;}
    else {return power(base, exponent - 1) * base;}
}

float ReLU(float x){
    if (x > 0){return x;}
    else {return 0.0;}
}

float ReLU_prime(float x){
    if (x > 0){return 1.0;}
    else {return 0.0;}
}

class ShapeException: std::exception {
    private:
        std::string msg;
    public:
        ShapeException (std::string m){msg = m;}
        char* what(){
            return msg.data();
        }
};

class Matrix {
    public:
        int rows;
        int cols;
        float* data;

        Matrix fromTable(int r, int c, float* data){
            Matrix M = Matrix(r, c);

            for (int j = 0; j < M.cols; j++){
                for (int i = 0; i < M.rows; i++){
                    M.setElement(
                        i,
                        j,
                        data[i * M.cols + j]
                    );
                }
            }

            return M;
        }

        Matrix(){
            rows = 0;
            cols = 0;
        }

        Matrix(int r, int c, float init) {
            rows = r;
            cols = c;

            data = (float*) malloc(sizeof(float) * rows * cols);

            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    data[i * cols + j] = init;
                }
            }
        }

        Matrix(int r, int c) {
            rows = r;
            cols = c;

            data = (float*) malloc(sizeof(float) * rows * cols);

            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    data[i * cols + j] = 0.0;
                }
            }
        }

        void diagonal(){
            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    this->setElement(i,j,i+j);
                }
            }
            
        }

        float getElement(int row, int col){
            return data[row * cols + col];
        }

        void setElement(int row, int col, float value){
            data[row * cols + col] = value;
        }

        void print(){
            for (int i = 0; i < rows; i++){
                std::cout << "|";
                for (int j = 0; j < cols; j++){
                    std::cout << " " << data[i * cols + j] << " ";
                }
                std::cout << "|\n";
            }
        }

        Matrix add(Matrix other){
            Matrix M = Matrix(rows, cols, 0.0);
            if (other.cols == 1){
                for (int i = 0; i < rows; i++){
                    for (int j = 0; j < cols; j++){
                        M.setElement(
                            i,
                            j,
                            this->getElement(i, j) + other.getElement(i, 0)
                        );
                    }
                }
            } else {
                for (int i = 0; i < rows; i++){
                    for (int j = 0; j < cols; j++){
                        M.setElement(
                            i,
                            j,
                            this->getElement(i, j) + other.getElement(i,j)
                        );
                    }
                }
            }
            return M;  
        }

        Matrix sub(Matrix other){
            Matrix M = Matrix(rows, cols, 0.0);
            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    M.setElement(
                        i,
                        j,
                        this->getElement(i, j) + other.getElement(i,j)
                    );
                }
            }
            return M;
        }

        Matrix scale(float scalar){
            Matrix M = Matrix(rows, cols, 0.0);
            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    M.setElement(
                        i,
                        j,
                        this->getElement(i, j) * scalar
                    );
                }
            }
            return M;
        }

        Matrix matmul(Matrix other){
            int k = rows;
            int m = other.cols;

            int l = other.rows;

            Matrix M = Matrix(k, m, 0.0);

            for (int i = 0; i < M.rows; i++){
                for (int j = 0; j < M.cols; j++){
                    float element = 0.0;
                    for (int x = 0; x < l; x++){
                        element += this->getElement(i, x) * other.getElement(x, j);
                    }
                    M.setElement(
                        i,
                        j,
                        element
                    );
                }
            }
            return M;
        }

        Matrix T(){
            Matrix M = Matrix(cols, rows, 0.0);
            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    M.setElement(
                        j,
                        i,
                        this->getElement(i, j)
                    );
                }
            }
            return M;
        }

        Matrix map(float (*func)(float)){
            Matrix M = Matrix(rows, cols);
            for (int i = 0; i < rows; i++){
                for (int j = 0; j < rows; j){
                    M.setElement(
                        i,
                        j,
                        func(this->getElement(i, j))
                    );
                }
            }
            return M;
        }

        Matrix sum(int axis){
            if (axis == 0){
                Matrix M = Matrix(1, cols);
                for (int j = 0; j < cols; j++){
                    float element = 0.0;
                    for (int i = 0; i < rows; i++){
                        element += this->getElement(i, j);
                    } 
                    M.setElement(0, j, element);
                }
                return M;
            } else {
                Matrix M = Matrix(rows, 1);
                for (int i = 0; i < rows; i++){
                    float element = 0.0;
                    for (int j = 0; j < rows; j++){
                        element += this->getElement(i, j);
                    } 
                    M.setElement(i, 0, element);
                }
                return M;
            }
        }

        Matrix hadamard(Matrix other){
            Matrix M = Matrix(other.rows, other.cols);
            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    M.setElement(
                        i,
                        j,
                        this->getElement(i,j) * other.getElement(i,j)
                    );
                }
            }
            return M;
        }

        Matrix concat(Matrix other){
            Matrix M = Matrix(rows, cols + other.cols);
            for (int i = 0; i < rows; i++){
                for (int j = 0; j < cols; j++){
                    M.setElement(
                        i,
                        j,
                        this->getElement(i, j)
                    );
                }
            }

            for (int i = 0; i < rows; i++){
                for (int j = 0; j < other.cols; j++){
                    M.setElement(
                        i, j + cols, other.getElement(i, j)
                    );
                }
            }
            return M;
        }
};

struct ForwardPropagationResult {
    Matrix* activated;
    Matrix* unactivated;
};

class Network{
    public:
        int* shape;
        int layers;

        Matrix* weights;
        Matrix* biases;
        float alpha;

        Network(int layers, int* shape){
            weights = (Matrix*) malloc(sizeof(Matrix) * (layers - 1));
            biases  = (Matrix*) malloc(sizeof(Matrix) * (layers - 1));

            for (int index = 0; index < layers - 1; index++){
                weights[index] = Matrix(shape[index + 1], shape[index]);
                biases[index] = Matrix(shape[index + 1], 1);
            }
        }

        ForwardPropagationResult feed(Matrix I){
            Matrix activated[layers];
            Matrix unactiavted[layers];

            activated[0] = I;
            unactiavted[0] = I;

            Matrix A = I;
            for (int index = 0; index < layers - 1; index++){
                Matrix U = (weights[index].matmul(A)).add(biases[index]);
                A = U.map(&ReLU);

                activated[index + 1] = A;
                unactiavted[index + 1] = U; 
            }

            ForwardPropagationResult R;
            R.activated = activated;
            R.unactivated = unactiavted;
            return R;
        }

        Matrix forward(Matrix I){
            ForwardPropagationResult R = this->feed(I);

            return R.activated[layers - 1];
        }

        void back(Matrix I, Matrix O){
            ForwardPropagationResult R = this->feed(I);

            Matrix dZ = R.activated[layers - 1].sub(O);
            for (int index = layers - 1; index > 0; index--){
                int m = R.activated[index].cols;

                Matrix dW = dZ.matmul(R.activated[index - 1].T()).scale(1.0/float(m));
                Matrix dB = dZ.sum(1).scale(1.0/float(m));

                dZ = weights[index - 1].T().matmul(dZ).hadamard(R.unactivated[index - 1].map(&ReLU_prime));

                weights[index - 1] = weights[index - 1].sub(dW.scale(alpha));
                biases[index - 1] = biases[index - 1].sub(dB.scale(alpha));
            }
        }
};

Matrix convert(int number, int size){
    Matrix M = Matrix(size, 1);
    for (int index = 0; index < size; index++){
        float bit = number % 2;
        number    = number / 2;
        M.setElement(
            index,
            0,
            bit
        );
    }

    return M;
}

Matrix oneHot(int place, int size){
    Matrix M = Matrix(size, 1);
    for (int index = 0; index < size; index++){
        if (index == place){
            M.setElement(
                index,
                0,
                1.0
            );
        } else {
            M.setElement(
                index,
                0,
                0.0
            );
        }
    }
    return M;
}

struct DataB {
    Matrix* d;
    Matrix* r;
};

DataB generate(int size){
    Matrix* data = (Matrix*) malloc(sizeof(Matrix) * power(2, size));
    Matrix* resu = (Matrix*) malloc(sizeof(Matrix) * power(2, size));
    for (int index = 0; index < power(2, size); index++){
        data[index] = convert(index, size);
        if (index % 7 == 0){
            resu[index] = oneHot(0, 2);
        } else {
            resu[index] = oneHot(1, 2);
        }
    }

    Matrix* datab = (Matrix*) malloc(sizeof(Matrix) * power(2, size - 3));
    Matrix* resub = (Matrix*) malloc(sizeof(Matrix) * power(2, size - 3));
    for (int i = 0; i < power(2, size - 3); i++){
        Matrix d = Matrix(size, 0);
        Matrix r = Matrix(2, 0);
        for (int j = 0; j < 8; j++){
            d = d.concat(data[i * 8 + j]);
            r = r.concat(resu[i * 8 + j]);
        }
        datab[i] = d;
        resub[i] = r;
    }

    DataB d;
    d.d = datab;
    d.r = resub;

    return d;
}

int main(){
    DataB data = generate(8);

    int layers = 4;
    int* shape = (int*) malloc(4 * layers);
    shape[0] = 8;
    shape[1] = 20;
    shape[2] = 10;
    shape[3] = 2;
    Network NN = Network(layers, shape);
    for (int i = 0; i < 100; i++){
        std::cout << i << std::endl;
        for (int block = 0; block < power(2, 8 - 3); block++){
            NN.back(data.d[block], data.r[block]);
        }
    }

    Matrix r = NN.forward(data.d[0]);

    data.d[0].print();
    std::cout << std::endl;

    data.r[0].print();
    std::cout << std::endl;

    r.print();
    std::cout << std::endl;
}

File: python/pure.py
import random
import struct

class Matrix:

    @classmethod
    def fromArray(cls, rows: int, columns: int, data: list[float]):
        M = cls(rows, columns)

        for i in range(rows):
            for j in range(columns):
                M[i, j] = data[i * columns + j]
        
        return M
    
    @classmethod
    def fromTable(cls, table: list[list[tuple]]):
        rows = len(table)
        cols = len(table[0])

        print(table)
        M = Matrix(rows = rows, columns = cols, init = 0.0)

        for i in range(rows):
            for j in range(cols):
                M[i,j] = table[i][j]

        return M

    @classmethod
    def fromBytes(cls, rows: int, columns: int, data: bytearray):
        M = cls(rows, columns)

        for j in range(columns):
            for i in range(rows):
                M[i, j] = struct.unpack('f', data[:4])[0] 
                data = data[4:]
        return M

    @classmethod
    def random(cls, rows: int, columns: int):
        M = cls(rows, columns)

        for i in range(rows):
            for j in range(columns):
                M[i, j] = random.random() - .5

        return M

    def __init__(self, rows: int, columns: int, *, init: float = 0.0):
        self.rows    = rows
        self.columns = columns
        self.__data  = [init for _ in range(self.rows * self.columns)]

    def __getitem__(self, index):
        i, j = index
        return self.__data[i * self.columns + j]
    
    def __setitem__(self, index, value):
        i, j = index
        self.__data[i * self.columns + j] = value

    def __repr__(self) -> str:
        output = ''
        for i in range(self.rows):
            output += '|'
            for j in range(self.columns):
                output += f' {self[i,j]:8.5f} '
            output += '|\n'
        return output

    def __add__(U, V):
        if U.rows != V.rows: raise IndexError('Cannot add two matrixes')
        if V.columns == 1:
            M = Matrix(rows = U.rows, columns = U.columns, init = 0.0)
            for i in range(U.rows):
                for j in range(U.columns):
                    M[i,j] = U[i,j] + V[i, 0]
        else:
            M = Matrix(rows = V.rows, columns = U.columns, init = 0.0)
            for i in range(U.rows):
                for j in range(V.columns):
                    M[i, j] = U[i, j] + V[i, j]
        return M
    
    def __sub__(U, V):
        if U.rows != V.rows or U.columns != V.columns: raise IndexError('Matrix sizes are different')

        M = Matrix(rows = V.rows, columns = U.columns, init = 1.0)
        for i in range(U.rows):
            for j in range(V.columns):
                M[i, j] = U[i, j] - V[i, j]
        
        return M       

    def __mul__(U, c):        
        if type(c) == Matrix:
            M = Matrix(rows = U.rows, columns = U.columns)

            for i in range(U.rows):
                for j in range(U.columns):
                    M[i,j] = U[i,j] * c[i,j]
            return M
        else:
            M = Matrix(rows = U.rows, columns = U.columns)

            for i in range(U.rows):
                for j in range(U.columns):
                    M[i,j] = U[i,j] * c

            return M

    def __rmul__(U,c): return U.__mul__(c)

    def __matmul__(U, V):
        if U.columns != V.rows: raise IndexError('Sizes doesn\'t match')

        k, l, l , m = U.rows, U.columns, V.rows, V.columns

        M = Matrix(rows = k, columns = m)
        
        for i in range(M.rows):
            for j in range(M.columns):
                M[i, j] = sum(
                    U[i, x] * V[x, j] for x in range(l)
                )

        return M

    @property
    def T(self):
        M = Matrix(rows = self.columns, columns = self.rows)

        for i in range(self.rows):
            for j in range(self.columns):
                M[j, i] = self[i, j]

        return M

    def map(self, func):
        M = Matrix(rows = self.rows, columns = self.columns)

        for i in range(self.rows):
            for j in range(self.columns):
                M[i,j] = func(self[i, j])
        
        return M

    def serialize(self):
        buffer = bytearray()

        for j in range(self.columns):
            for i in range(self.rows):
                buffer += struct.pack('f', self[i,j])
        return buffer

    def sum(self, axis: int = 0):

        if axis == 0:
            M = Matrix(rows = 1, columns = self.columns, init = 0.0)
            for j in range(self.columns):
                M[0, j] = sum(self[i, j] for i in range(self.rows))
        elif axis == 1:
            M = Matrix(rows = self.rows, columns = 1, init = 0.0)
            for i in range(self.rows):
                M[i, 0] = sum(self[i, j] for j in range(self.columns))
        
        return M

    
class Network:

    @classmethod
    def getFromFile(cls, path: str):
        with open(path, 'rb') as file: data = bytearray(file.read())

        shape: list[int] = [struct.unpack('i', data[:4])[0],]
        while shape[-1]:
            data = data[4:]
            shape += struct.unpack('i', data[:4])

        data = data[4:]
        shape = shape[:-1]

        weights = []
        for rows, cols in zip(shape, shape[1:]):
            weights.append(
                Matrix.fromBytes(cols, rows, data[:4 * rows * cols])
            )
            data = data[4 * rows * cols:]

        vectors = []
        for cols in shape[1:]:
            vectors.append(
                Matrix.fromBytes(cols, 1, data[:4 * cols])
            )
            data = data[4 * cols:]

        n = cls(tuple(shape))
        n.weights = weights
        n.biases = vectors

        return n
    
    @staticmethod
    def ReLU(X: Matrix) -> Matrix: return X.map(lambda x: max(0, x))

    @staticmethod
    def ReLU_prime(X: Matrix) -> Matrix: return X.map(lambda x: x > 0)

    def __init__(self, shape: tuple[int, ...]):
        self.shape = shape

        self.weights: list[Matrix] = []
        self.biases: list[Matrix] = []
        self.alpha: float = 0.05

    def populate(self):
        for prev, next in zip(self.shape, self.shape[1:]):
            w = Matrix.random(next, prev)
            self.weights.append(w)

            b = Matrix.random(next, 1)
            self.biases.append(b)

    def __forward(self, I: Matrix) -> tuple[list[Matrix], list[Matrix]]:
        activated = [I,]
        unactivated = [I,]

        A = I
        for index in range(len(self.shape) - 1):
            U = self.weights[index] @ A + self.biases[index]
            A = self.ReLU(U)

            activated.append(A)
            unactivated.append(U)

        return activated, unactivated
    
    def forward(self, I: Matrix) -> Matrix:
        A, _ = self.__forward(I)
        return A[-1]
    
    def back(self, I: Matrix, O: Matrix):
        A, U = self.__forward(I)
        dZ = A[-1] - O
        for index in range(len(self.shape) - 1, 0, -1):
            m = A[index].columns

            dW = 1/m * dZ @ A[index - 1].T
            dB = 1/m * dZ.sum(axis = 1)

            dZ = self.weights[index - 1].T @ dZ * self.ReLU_prime(U[index - 1])

            self.weights[index - 1] = self.weights[index - 1] - self.alpha * dW
            self.biases[index - 1] = self.biases[index - 1] - self.alpha * dB

    def serialize(self):
        buffer = bytearray()

        for size in self.shape:
            buffer += struct.pack('i', size)
        buffer += struct.pack('i', 0x00)

        for matrix in self.weights:
            buffer += matrix.serialize()

        for vector in self.biases:
            buffer += vector.serialize()

        return buffer
    
    def putToFile(self, path: str):
        with open(path, 'wb') as file: file.write(self.serialize())

def convert(n: int, *, c: int = 8):
    bits = []
    for _ in range(c): n, r = divmod(n, 2); bits += [float(r),]
    return bits

def generate(size: int, bsize: int = 10):
    data = []
    for index in range(2 ** size):
        if index % 9 == 0:
            data.append(
                [convert(index, c=size), [1.0, 0.0]]
            )
        else:
            data.append(
                [convert(index, c=size), [0.0, 1.0]]
            )
    
    random.shuffle(data)

    batches = []
    for index in range(bsize):
        batches.append(
            data[index::bsize]
        )

    pairs = []
    for b in batches:
        Im = []
        Om = []
        for I, O in b:
            Im += [I,]
            Om += [O,]

        pairs.append(
            [Matrix.fromTable(Im).T, Matrix.fromTable(Om).T]
        )

    return pairs

if __name__ == '__main__':
    n = Network((8, 20, 10, 2))
    n.populate()

    data = generate(8, 20)

    I, O = data[1]
    E = n.forward(I)

    print(f'Input: \n{I}')
    print(f'Target: \n{O}')
    print(f'Result: \n{E}')
    print(f'Diff:\n{O - E}')
    print(f'Accuracy:\n{(O - E).sum(axis=1) * (1 / E.columns)}')
    print()
    for index in range(100):
        print(f'Iteration: {index}')
        for I, O in data:
            n.back(I, O)

    I, O = data[1]
    E = n.forward(I)

    print(f'Input: \n{I}')
    print(f'Target: \n{O}')
    print(f'Result: \n{E}')
    print(f'Diff:\n{O - E}')
    print(f'Accuracy:\n{(O - E).sum(axis=1) * (1 / E.columns)}')
    print()

    n.putToFile('test.lbt')

    n = Network.getFromFile('test.lbt')
    I, O = data[1]
    E = n.forward(I)

    print(f'Input: \n{I}')
    print(f'Target: \n{O}')
    print(f'Result: \n{E}')
    print(f'Diff:\n{O - E}')
    print(f'Accuracy:\n{(O - E).sum(axis=1) * (1 / E.columns)}')
    print()